### PR TITLE
TST: Groupby first/last/nth nan column test

### LINKED ISSUE #33591: BUG: groupby first()/last() change datatype of all NaN columns to float; nth() preserves datatype
- [x ] I have checked that this issue has not already been reported.
- [x ] I have confirmed this bug exists on the latest version of pandas.  
Version 1.0.3 from the conda distro.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
print('pandas version is', pd.__version__, '\n')

df = pd.DataFrame({'id':['a','b','b','c'], 'sym':['ibm','msft','msft','goog'], 'idx':range(4), 'sectype':['E','E','E','E']})
df['osi'] = df.sym.where(df.sectype=='O')  # add a column of NaNs that are object/str
print(df)
print(df.info())

# .nth(-1) does the right thing
x = df.groupby('id').nth(-1)
print('\nnth(-1)')
print(x.info())

# .last() converts the osi col to floats
x = df.groupby('id').last()
print('\nlast()')
print(x.info())

# .nth(0) does the right thing
x = df.groupby('id').nth(-1)
print('\nnth(0)')
print(x.info())

# .first() converts the osi col to floats
x = df.groupby('id').first()
print('\nfirst()')
print(x.info())

```

#### Problem description
Given a dataframe with an all-NaN column of str/objects, performing a groupby().first() will convert the all-NaN column to float.  This is true for .last() as well, but not for .nth(0) and .nth(-1), so these are workarounds.

This is a problem for me as the groupby().first() is in general code that is iteratively called and the results either pd.concat'd (resulting in a column with mixed types) or appended to an hdf file (causing failure on the write).

#### Expected Output
The resulting dataframe from a groupby().first()/.last() should have the same metadata (column structure and datatypes) as the input dataframe.  The result of .first()/.last() should match that of .nth(0)/.nth(-1) .

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 85 Stepping 4, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.0.3
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : 0.29.15
pytest           : 5.4.1
hypothesis       : 5.8.3
sphinx           : 2.4.4
blosc            : None
feather          : None
xlsxwriter       : 1.2.8
lxml.etree       : 4.5.0
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.13.0
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : 1.3.2
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.5.0
matplotlib       : 3.1.3
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.3
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : 5.4.1
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : 1.3.15
tables           : 3.6.1
tabulate         : None
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.8
numba            : 0.48.0
</details>

### PR DESCRIPTION
**Pull request by jamescobonkerr**
- [X] xref https://github.com/pandas-dev/pandas/issues/33591#issuecomment-615365647
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

The first two asserts (`first()`/`last()`) fail on `1.0.3`, as expected.

### CODE REVIEW

File: pandas/tests/groupby/test_nth.py
Comment (WillAyd) [inline pandas/tests/groupby/test_nth.py:146]: You can just add this column to the constructor

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks - just a minor nit to make things more readable otherwise lgtm

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @jamescobonkerr for the PR. does this test follow the same code path as the code sample in the issue OP?, see https://github.com/pandas-dev/pandas/issues/33591#issuecomment-616164785

**Comment by simonjayhawkins** [comment]
> The first two asserts (`first()`/`last()`) fail on `1.0.3`, as expected

this was changed in #32949

be86b6583f97a49ded14bdd4290ccf721bf1f645 is the first new commit
commit be86b6583f97a49ded14bdd4290ccf721bf1f645
Author: jbrockmendel <jbrockmendel@gmail.com>
Date:   Fri Apr 10 09:18:06 2020 -0700

    CLN: remove unnecessary Series._convert calls (#32949)

**Comment by jamescobonkerr** [comment]
> Thanks @jamescobonkerr for the PR. does this test follow the same code path as the code sample in the issue OP?, see [#33591 (comment)](https://github.com/pandas-dev/pandas/issues/33591#issuecomment-616164785)

Ah, you're right — pardon the oversight.



> > The first two asserts (`first()`/`last()`) fail on `1.0.3`, as expected
> 
> this was changed in #32949
> 
> [be86b65](https://github.com/pandas-dev/pandas/commit/be86b6583f97a49ded14bdd4290ccf721bf1f645) is the first new commit
> commit [be86b65](https://github.com/pandas-dev/pandas/commit/be86b6583f97a49ded14bdd4290ccf721bf1f645)
> Author: jbrockmendel [jbrockmendel@gmail.com](mailto:jbrockmendel@gmail.com)
> Date: Fri Apr 10 09:18:06 2020 -0700
> 
> ```
> CLN: remove unnecessary Series._convert calls (#32949)
> ```

Removing the `_convert` call from `core.groupby.generic.DataFrameGroupBy._wrap_aggregated_output` seems to fix the reported issue, however, breaks a range of other tests. I don't have a great understanding of how `GroupBy` works. I'll close this PR.

**Comment by simonjayhawkins** [comment]
@jamescobonkerr Thanks for working on this.

The test you have added here does have value since it is testing a change in behaviour against regression.

As you pointed out this test fails on 1.0.3 and is fixed on master. AFAICT this is working since #32949

I'll reopen this PR, but once merged will not close #33591 (rather than open a new issue at this point, i'll reference you code sample in #33591)

**Review by simonjayhawkins — APPROVED** [review]
Thanks @jamescobonkerr lgtm pending green ci.

File: pandas/tests/groupby/test_nth.py
Comment (jreback) [inline pandas/tests/groupby/test_nth.py:149]: construct an expected frame and use
tm.assert_frame_equal(result, expected)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_nth.py
Comment (jreback) [inline pandas/tests/groupby/test_nth.py:151]: dont' we need to *also* tests `.nth(-1)`?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jamescobonkerr always happy to take these

**Pull request merged**
